pkg/kubelet/cm/deviceplugin/device_plugin_handler.go

```diff
@@ -0,0 +1,289 @@
+	return h.devicePluginManager.Devices()
+}
+
+// Returns list of device Ids we need to allocate with Allocate rpc call.
+// Returns empty list in case we don't need to issue the Allocate rpc call.
+func (h *HandlerImpl) devicesToAllocate(podUID, contName, resource string, required int) (sets.String, error) {
+	h.Lock()
+	defer h.Unlock()
+	needed := required
+	// Gets list of devices that have already been allocated.
+	// This can happen if a container restarts for example.
+	devices := h.podDevices.containerDevices(podUID, contName, resource).Union(sets.NewString())
+	if devices != nil {
+		glog.V(3).Infof("Found pre-allocated devices for resource %s container %q in Pod %q: %v", resource, contName, podUID, devices.List())
+		needed = needed - devices.Len()
+	}
+	if needed == 0 {
+		// No change, no work.
+		return nil, nil
+	}
+	if needed < 0 {
+		// Pod is downsizing its resource requests. We need to make Allocate rpc call
+		// again to re-generate device runtime config. allocatedDevices and podDevices
+		// will be updated after Allocate rpc call succeeds. If the rpc call fails, we
+		// will keep the current cached state unchanged.
+		for devices.Len() > required {
+			devices.PopAny()
+		}
+		glog.V(3).Infof("Return downsized devices %v", devices.List())
+		return devices, nil
+	}
+	// Needs to allocate additional devices.
+	if h.allocatedDevices[resource] == nil {
+		h.allocatedDevices[resource] = sets.NewString()
+	}
+	// Gets Devices in use.
+	devicesInUse := h.allocatedDevices[resource]
+	// Gets a list of available devices.
+	available := h.allDevices[resource].Difference(devicesInUse)
+	if int(available.Len()) < needed {
+		return nil, fmt.Errorf("requested number of devices unavailable for %s. Requested: %d, Available: %d", resource, needed, available.Len())
+	}
+	allocated := available.UnsortedList()[:needed]
+	// Updates h.allocatedDevices with allocated devices to prevent them
+	// from being allocated to other pods/containers, given that we are
+	// not holding lock during the rpc call.
+	for _, device := range allocated {
+		h.allocatedDevices[resource].Insert(device)
+		devices.Insert(device)
+	}
+	return devices, nil
+}
+
+// Allocate attempts to allocate all of required extended resources for
+// the input container, issues an Allocate rpc request for each of such
+// resources, processes their AllocateResponses, and updates the cached
+// containerDevices on success.
+func (h *HandlerImpl) Allocate(pod *v1.Pod, container *v1.Container, activePods []*v1.Pod) error {
+	h.updateAllocatedDevices(activePods)
+	podUID := string(pod.UID)
+	contName := container.Name
+	for k, v := range container.Resources.Limits {
+		resource := string(k)
+		needed := int(v.Value())
+		glog.V(3).Infof("needs %d %s", needed, resource)
+		if _, registeredResource := h.allDevices[resource]; !registeredResource {
+			continue
+		}
+		allocDevices, err := h.devicesToAllocate(podUID, contName, resource, needed)
+		if err != nil {
+			return err
```

Comment:
Will devices get stranded if Allocate fails? IIUC, devices are removed from Allocatable pool prior to actual allocation.

Reply:
nvm. I see the note on `updateAllocatedDevices` below